src/coreclr/src/System.Private.CoreLib/src/System/Threading/ProcessorIdCache.cs

```diff
@@ -0,0 +1,124 @@
+            // returns -1.  As a fallback in that case and to spread the threads across the buckets
+            // by default, we use the current managed thread ID as a proxy.
+            if (currentProcessorId < 0) currentProcessorId = Environment.CurrentManagedThreadId;
+
+            Debug.Assert(s_processorIdRefreshRate <= ProcessorIdCacheCountDownMask);
+
+            // Mask with int.MaxValue to ensure the execution Id is not negative
+            t_currentProcessorIdCache = ((currentProcessorId << ProcessorIdCacheShift) & int.MaxValue) | s_processorIdRefreshRate;
+
+            return currentProcessorId;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static int GetCurrentProcessorId()
+        {
+            int currentProcessorIdCache = t_currentProcessorIdCache--;
+            if ((currentProcessorIdCache & ProcessorIdCacheCountDownMask) == 0)
+            {
+                return RefreshCurrentProcessorId();
+            }
+
+            return currentProcessorIdCache >> ProcessorIdCacheShift;
+        }
+
+        // If GetCurrentProcessorNumber takes any nontrivial time (compared to TLS access), return false.
+        // Check more than once - to make sure it was not because TLS was delayed by GC or a context switch.
+        internal static unsafe int ProcessorNumberSpeedCheck()
+        {
+            // NOTE: We do not check the frequency of the Stopwatch.
+            //       The frequency often does not match the actual timer refresh rate anyways.
+            //       If the resolution, precision or access time to the timer are inadequate for our measures here,
+            //       the test will fail anyways.
+
+            // default values pretend to be very long samples with default ratio
+            double minID = Stopwatch.Frequency * 50;  // 50 sec
+            double minTLS = Stopwatch.Frequency;      // 1 sec
+
+            // warm up the code paths.
+            UninlinedThreadStatic();
+            if (Thread.GetCurrentProcessorNumber() < 0)
+                return MaxIdRefreshRate;
+
+            long oneMicrosecond = Stopwatch.Frequency / 1000000;
+            for (int i = 0; i < 10; i++)
+            {
+                // we will measure at least 16 iterations and at least 1 microsecond
+                int iters = 16;
+                long t1 = 0;
+                while (t1 < oneMicrosecond)
+                {
+                    iters *= 2;
+                    t1 = Stopwatch.GetTimestamp();
+                    for (int j = 0; j < iters; j++)
+                    {
+                        Thread.GetCurrentProcessorNumber();
+                    }
+                    t1 = Stopwatch.GetTimestamp() - t1;
+                }
+
+                minID = Math.Min(minID, (double)t1 / iters);
+
+                // we will measure at least 16 iterations and at least 1 microsecond
+                iters = 16;
```

Comment:
Keep iters from GetCurrentProcessorNumber measurement with assumption that TLS access is never slower than GetCurrentProcessorNumber?

Reply:
On Zen2 and recent ARM64 TLS is 2x slower. I think the gap cant be much more than that though, so starting with 1/2 of ProcNumber measurement seems reasonable.

---
In reply to: [356993885](https://github.com/dotnet/runtime/pull/467#discussion_r356993885) [](ancestors = 356993885)